sklearn/cluster/_hdbscan/_linkage.pyx

```diff
@@ -0,0 +1,245 @@
+# Minimum spanning tree single linkage implementation for hdbscan
+# Authors: Leland McInnes, Steve Astels
+# License: 3-clause BSD
+
+import numpy as np
+cimport numpy as cnp
+import cython
+
+from libc.float cimport DBL_MAX
+
+from sklearn.metrics._dist_metrics cimport DistanceMetric
+
+
+cpdef cnp.ndarray[cnp.double_t, ndim=2] mst_linkage_core(
+    cnp.ndarray[cnp.double_t, ndim=2] distance_matrix
+):
+
+    cdef:
+        cnp.ndarray[cnp.intp_t, ndim=1] node_labels
+        cnp.ndarray[cnp.intp_t, ndim=1] current_labels
+        cnp.ndarray[cnp.double_t, ndim=1] current_distances
+        cnp.ndarray[cnp.double_t, ndim=1] left
+        cnp.ndarray[cnp.double_t, ndim=1] right
+        cnp.ndarray[cnp.double_t, ndim=2] result
+
+        cnp.ndarray label_filter
+
+        cnp.intp_t current_node
+        cnp.intp_t new_node_index
+        cnp.intp_t new_node
+        cnp.intp_t i
+
+    result = np.zeros((distance_matrix.shape[0] - 1, 3))
+    node_labels = np.arange(distance_matrix.shape[0], dtype=np.intp)
+    current_node = 0
+    current_distances = np.infty * np.ones(distance_matrix.shape[0])
+    current_labels = node_labels
+    for i in range(1, node_labels.shape[0]):
+        label_filter = current_labels != current_node
+        current_labels = current_labels[label_filter]
+        left = current_distances[label_filter]
+        right = distance_matrix[current_node][current_labels]
+        current_distances = np.where(left < right, left, right)
+
+        new_node_index = np.argmin(current_distances)
+        new_node = current_labels[new_node_index]
+        result[i - 1, 0] = <double> current_node
+        result[i - 1, 1] = <double> new_node
+        result[i - 1, 2] = current_distances[new_node_index]
+        current_node = new_node
+
+    return result
+
+
+cpdef cnp.ndarray[cnp.double_t, ndim=2] mst_linkage_core_vector(
```

Comment:
and this one `mst_from_data_matrix`

Reply:
Is there a computational regime where Prim's algorithm is indeed interesting in place of Boruvska's algorithm? Otherwise, I would think that this part will become quickly obsolete.

Reply:
The original heuristic they included was to use prims when `X.shape[1] > 60`, however I have yet to empirically verify.

Reply:
I did some benchmarking once upon a time a long time ago between the two regimes and 60 was about the crossover point. I never revisited it, and it is very much just a heuristic.

The crossover comes because the dual tree Boruvka requires the spatial indexing trees to be efficient. In higher dimensions tree based spatial indexing tends toward brute force (due to too much backtracking being required). That means the crossover is in the dimension (and data distribution) dependent constants hidden in the big O notation. So the real question is when the spatial indexing tree performance fails to give enough gains. I think that is very hard to estimate well.